File: dijkstra.cpp
#include <iostream>
using namespace std;

const int MAX_VERTICES = 100;
const int INF = 1e9;

class Dijkstra {
private:
    int graph[MAX_VERTICES][MAX_VERTICES];
    int numVertices;

public:
    Dijkstra(int n) : numVertices(n) {
        // Initialize the graph with INF (representing absence of edge)
        for (int i = 0; i < numVertices; ++i) {
            for (int j = 0; j < numVertices; ++j) {
                graph[i][j] = INF;
            }
        }
    }

    void addEdge(int u, int v, int weight) {
        graph[u][v] = weight;
    }

    void dijkstraShortestPath(int source) {
        int distance[MAX_VERTICES];
        bool visited[MAX_VERTICES];

        // Initialize distances and visited array
        for (int i = 0; i < numVertices; ++i) {
            distance[i] = INF;
            visited[i] = false;
        }

        // Distance from the source to itself is 0
        distance[source] = 0;

        // Find shortest path for all vertices
        for (int count = 0; count < numVertices - 1; ++count) {
            int u = minDistance(distance, visited);
            visited[u] = true;

            for (int v = 0; v < numVertices; ++v) {
                if (!visited[v] && graph[u][v] != INF && distance[u] != INF &&
                    distance[u] + graph[u][v] < distance[v]) {
                    distance[v] = distance[u] + graph[u][v];
                }
            }
        }

        printShortestPaths(source, distance);
    }

    int minDistance(const int distance[], const bool visited[]) {
        int minIndex, minValue;
        minIndex = minValue = INF;

        for (int v = 0; v < numVertices; ++v) {
            if (!visited[v] && distance[v] < minValue) {
                minValue = distance[v];
                minIndex = v;
            }
        }

        return minIndex;
    }

    void printShortestPaths(int source, const int distance[]) {
        cout << "Shortest paths from source " << source << " to all other vertices:" << endl;
        for (int i = 0; i < numVertices; ++i) {
            cout << "Vertex " << i << ": ";
            if (distance[i] == INF) {
                cout << "INFINITY" << endl;
            } else {
                cout << distance[i] << endl;
            }
        }
    }
};

int main() {
    int numVertices = 6;

    Dijkstra dijkstra(numVertices);

    dijkstra.addEdge(0, 1, 5);
    dijkstra.addEdge(0, 2, 2);
    dijkstra.addEdge(1, 3, 4);
    dijkstra.addEdge(2, 1, 1);
    dijkstra.addEdge(2, 3, 7);
    dijkstra.addEdge(3, 4, 3);
    dijkstra.addEdge(4, 0, 2);
    dijkstra.addEdge(4, 5, 1);
    dijkstra.addEdge(5, 3, 6);

    int source = 0;

    dijkstra.dijkstraShortestPath(source);

    return 0;
}

File: Dijkstra.py
import sys

V = 9  # no. of vertices

def minDistance(dist, sptSet):
    min_val = sys.maxsize
    min_index = -1
    for v in range(V):
        if (not sptSet[v]) and (dist[v] <= min_val):
            min_val = dist[v]
            min_index = v
    return min_index

def printSolution(dist):
    print("Vertex \t Distance From Source")
    for i in range(V):
        print(i, " \t\t", dist[i])

def dijkstra(graph, src):
    min_val = sys.maxsize
    dist = [min_val] * V
    sptSet = [False] * V
    dist[src] = 0

    for count in range(V - 1):
        u = minDistance(dist, sptSet)
        sptSet[u] = True
        for v in range(V):
            if (not sptSet[v]) and (graph[u][v] != 0) and (dist[u] != min_val) and (dist[u] + graph[u][v] < dist[v]):
                dist[v] = dist[u] + graph[u][v]

    printSolution(dist)

if __name__ == "__main__":
    graph = [
        [0, 4, 0, 0, 0, 0, 0, 8, 0],
        [4, 0, 8, 0, 0, 0, 0, 11, 0],
        [0, 8, 0, 7, 0, 4, 0, 0, 2],
        [0, 0, 7, 0, 9, 14, 0, 0, 0],
        [0, 0, 0, 9, 0, 10, 0, 0, 0],
        [0, 0, 4, 14, 10, 0, 2, 0, 0],
        [0, 0, 0, 0, 0, 2, 0, 1, 6],
        [8, 11, 0, 0, 0, 0, 1, 0, 7],
        [0, 0, 2, 0, 0, 0, 6, 7, 0]
    ]
    dijkstra(graph, 0)
